operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java

```diff
@@ -246,4 +259,154 @@ public void testHostnameStrictDisabled() {
+
+            Log.info("Updating DB Secret, expecting restart");
+            testDeploymentRestarted(Set.of(kc1), Set.of(), () -> {
+                dbSecret.getData().put(UUID.randomUUID().toString(), "YmxhaGJsYWg=");
+                k8sclient.secrets().createOrReplace(dbSecret);
+            });
+
+            Log.info("Updating TLS Secret, expecting restart");
+            testDeploymentRestarted(Set.of(kc1), Set.of(), () -> {
+                tlsSecret.getData().put(UUID.randomUUID().toString(), "YmxhaGJsYWg=");
+                k8sclient.secrets().createOrReplace(tlsSecret);
+            });
+
+            Log.info("Updating DB Secret metadata, NOT expecting restart");
+            testDeploymentRestarted(Set.of(), Set.of(kc1), () -> {
+                dbSecret.getMetadata().getLabels().put(UUID.randomUUID().toString(), "YmxhaGJsYWg");
+                k8sclient.secrets().createOrReplace(dbSecret);
+            });
+
+            Log.info("Deploying second KC");
+            deployKeycloak(k8sclient, kc2, true);
+
+            Log.info("Updating DB Secret, expecting restart of both KCs");
+            testDeploymentRestarted(Set.of(kc1, kc2), Set.of(), () -> {
+                dbSecret.getData().put(UUID.randomUUID().toString(), "YmxhaGJsYWg=");
+                k8sclient.secrets().createOrReplace(dbSecret);
+            });
+
+            Log.info("Updating KC1 to not to rely on DB Secret");
+            kc1.getSpec().getServerConfiguration().put("KC_DB_USERNAME", new ValueOrSecret("postgres"));
+            kc1.getSpec().getServerConfiguration().put("KC_DB_PASSWORD", new ValueOrSecret("testpassword"));
+            testDeploymentRestarted(Set.of(kc1), Set.of(kc2), () -> {
+                deployKeycloak(k8sclient, kc1, false, false);
+            });
+
+            Log.info("Updating DB Secret, expecting restart of just KC2");
+            testDeploymentRestarted(Set.of(kc2), Set.of(kc1), () -> {
+                dbSecret.getData().put(UUID.randomUUID().toString(), "YmxhaGJsYWg=");
+                k8sclient.secrets().createOrReplace(dbSecret);
+            });
+
+            Log.info("Deleting KC2");
+            k8sclient.resources(Keycloak.class).inNamespace(namespace).delete(kc2);
+            Awaitility.await()
+                    .untilAsserted(() -> assertThat(k8sclient.apps().deployments()
+                            .inNamespace(namespace)
+                            .withName(kc2.getMetadata().getName())
+                            .get()).isNull());
+
+            Log.info("Trying to update DB Secret that is no longer needed by any CR");
+            testDeploymentRestarted(Set.of(), Set.of(kc1), () -> {
+                dbSecret.getData().put(UUID.randomUUID().toString(), "YmxhaGJsYWg=");
+                k8sclient.secrets().createOrReplace(dbSecret);
+            });
+
+            Awaitility.await().untilAsserted(() -> {
```

Comment:
This condition doesn't depend on the action right above, e.g. this should already be the case without the last DB secret update.

Reply:
It does depend on it. Since we're not using finalizers and the logic for cleaning up the labels is "hidden" in the event source, we need to modify the Secret to trigger the clean up.